File: Java/54.java
import java.util.LinkedList;
import java.util.List;

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> result = new LinkedList<>();
        if (matrix == null || matrix.length == 0) {

        }
        int n = matrix.length, m = matrix[0].length;
        int up = 0, down = n - 1;
        int left = 0, right = m - 1;
        while (result.size() < n * m) {
            for (int j = left; j <= right && result.size() < n * m; j++) {
                result.add(matrix[up][j]);
            }

            for (int i = up + 1; i <= down - 1 && result.size() < n * m; i++) {
                result.add(matrix[i][right]);
            }
            for (int j = right; j >= left && result.size() < n * m; j--) {
                result.add(matrix[down][j]);

            }
            for (int i = down - 1; i >= up + 1 && result.size() < n * m; i--) {
                result.add(matrix[i][left]);
            }

            left++;
            right--;
            up++;
            down--;
        }
        return result;
    }
}

File: Python/54.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        n, m = len(matrix), len(matrix[0])
        nm = n * m
        result = []
        top, bottom, left, right = 0, n - 1, 0, m - 1

        while len(result) < nm:
            for i in range(left, right + 1):
                result.append(matrix[top][i])
            top += 1

            for i in range(top, bottom + 1):
                result.append(matrix[i][right])
            right -= 1

            if top <= bottom:
                for i in range(right, left - 1, -1):
                    result.append(matrix[bottom][i])
                bottom -= 1

            if left <= right:
                for i in range(bottom, top - 1, -1):
                    result.append(matrix[i][left])
                left += 1

        return result
